File: src/cpp/subscriber.cpp
#include "subscriber.h"

#include <thread>
#include <chrono>

#include <zmq.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/filesystem.hpp>

#include <sys/stat.h>
#include <libgen.h>
#include <stdio.h>

#include "error.h"

using namespace bjcomm;
using namespace boost::algorithm;
using namespace boost::filesystem;

Subscriber::Subscriber(std::string address): 
    _address(address) {}
    
bool Subscriber::start(){
    //if already started return
    if(_running) return false;
    
    _socket = new zmq::socket_t (_context, ZMQ_SUB);
    
    //try to bind socket and set options
    try{
        //create path to directory if does not exists
        path comm_path(BJCOMM_COMMON_PATH+_address);
        create_directories(comm_path.parent_path());
        
        //bind socket
        _socket->bind(("ipc://"+comm_path.string()).c_str());
        
        //try to set permissions
        try{
            permissions(BJCOMM_COMMON_PATH, perms::owner_all | perms::group_all | perms::others_all);
            permissions(comm_path.parent_path(), perms::owner_all | perms::group_all | perms::others_all);
            permissions(comm_path, perms::owner_all | perms::group_all | perms::others_all);
        }catch(filesystem_error){}

        //FIXME: bind to all sockets, later add possibility to define message types and ignore others
        _socket->setsockopt(ZMQ_SUBSCRIBE, "", 0);
    }catch(zmq::error_t &e){
        delete _socket;
        _socket = 0;
        
        return false;
    }
    
    _running = true;
    return true;
}

Message Subscriber::receive(){
    //throw exception if trying to receive message on closed channel
    if(!_running) throw CommunicationError("Trying to receive message on channel that is not running!");
        
    //block read message
    zmq::message_t zmq_msg;
    _socket->recv(&zmq_msg);
        
    //read type and data
    std::string str(static_cast<char*>(zmq_msg.data()), zmq_msg.size());
    trim(str);
    
    size_t pos = str.find(' ');
    std::string type;
    std::string data;
    if(pos == std::string::npos){
        type = str;
    }else{
        type = str.substr(0, pos);
        data = str.substr(pos);
        trim(data);
    }
    Message msg(type, data);
    
    return msg;
}

File: src/java/Subscriber.java
/*
 * Copyright (c) 2015 Blue Jay Eindhoven
 */
package nl.bluejayeindhoven.bjcomm;

import java.io.File;
import java.nio.charset.StandardCharsets;

import org.zeromq.ZMQ;
import org.zeromq.ZMQException;

import nl.bluejayeindhoven.bjcomm.Message;
import nl.bluejayeindhoven.bjcomm.CommunicationError;

/**
 * Interface that subscribes to all messages on a socket it binds
 *
 * @see Publisher
 * @author Koen Wolters
 */
public class Subscriber extends CommunicationInterface{
    /**
     * Address of the socket
     */
    private String address;
    
    /**
     * Constructs a subscriber that shall communicate over the supplied address
     *
     * @param address address of the socket
     */
    public Subscriber(String address){
        this.address = COMMON_PATH+address;
    }
    
    /**
     * {@inheritDoc}
     *
     * Creates a socket, bind the address and subscribe to all channels
     */
    public boolean start(){
        //if already started return false (ALERT: should not happen)
        if(running) return false;
        
        socket = context.socket(ZMQ.SUB);
        
        try{
            //create file if not exists
            new File(COMMON_PATH+address).mkdirs();
        
            //bind socket
            socket.bind("ipc://"+COMMON_PATH+address);
            
            //FIXME: currently binding to all sockets
            socket.subscribe(new byte[0]);
        }catch(ZMQException e){
            socket.close();
            socket = null;
            return false;
        }
        
        running = true;
        return true;
    }
    
    /**
     * Receive a message over the channel. Blocks until message is received.
     *
     * @return msg message that is received on the channel
     *
     * @throws CommunicationError if interface is not started
     */
    public Message receive(){
        //throw exception if trying to receive message on closed channel
        if(!running) throw new CommunicationError("Trying to receive message on channel that is not running!");
        
        //read type and data
        String str = socket.recvStr(StandardCharsets.US_ASCII);
        str = str.trim();
        
        String[] parts = str.split(" ", 2);
        
        String type;
        String data;
        if(parts.length < 2){
            type = parts[0];
            data = "";
        }else{
            type = parts[0];
            data = parts[1];
            data = data.trim();
        }
        Message msg = new Message(type, data);
        
        return msg;
    }
}